README
fonty-rg http://kassiopeia.juls.savba.sk/~garabik/softwarefonty-rg
version: 0.6
by Radovan Garabík <garabik @ kassiopeia.juls.savba.sk>
License: GPL


This package is heavily based on fonty package by Marcin 'Qrczak' Kowalczyk
(http://kki.net.pl/qrczak/programy/linux/fonty/).
Some glyphs are obtained from GNU unifont by David Starner
(http://dvdeug.dhis.org/unifont.html), originally by Roman Czyborra
(http://www.czyborra.com).

To build fonts: run build.sh script.
To modify fonts: edit files in source/ directory.
To add new charsets: add *.txt file into charsets/ directory.

To use fonts: there are two small scripts provided, iso and utf8.
iso takes as an argument number of ISO-8859 encoding, switching 
console font into font for that encoding, and utf8 switches console
into Unicode mode, using big Unicode font. See manpages
for details.

About fonts:

LatCyrGr-16.psf is font containing most of ISO Latin characters,
lot of cyrillic chars, lot of greek chars, and some punctuation.
It is good for Unicode linux console. This is 512 character
font, means that you lose bold colour on linux console, unless
you are using framebuffer.

chavo.psf is font containing lot of ISO-8859-2 repertoire,
some additional ISO-8859-1 characters, Russian cyrillic
characters and special Esperanto characters. It was
specifically created to cover Czech, Slovak, Polish,
Hungarian, Russian, German, Esperanto. Of course, it accidentally
covers much wider range :-)
Good as a console font if you want to retain bold colours
(and do not need some ISO-8859-1 characters - there was
simply not enough space for them).

uni-compact-16.sbf.gz - compact Unicode source font -
good for use with dynafont (part of fonty package).


about glyphs:
glyphs from Basic Latin, Latin-1 Supplement, Latin Extended-A,
as well as most often used Cyrillic and Greek are verified, with
consistent typography. Characters from other ranges need some work
- mostly shifting by one or two pixels, and redrawing them with
consistent width.

For an excellent introduction into font internals, 
read Documentation.html (written by Erika Pacholleck <pacholleck.e @ gmx.de>)


                               Radovan Garabík



debian/changelog
fonty-rg (0.7-1) unstable; urgency=low

  * rewrite iso and utf8 scripts using kbd,
    not console-tools (closes: #863065)
  * change into non-native package

 -- Radovan Garabík <garabik@kassiopeia.juls.savba.sk>  Mon, 29 May 2017 19:13:14 +0200

fonty-rg (0.6) unstable; urgency=low

  * do not save gzip timestamp, allowing reproducible build (closes: #777061)
  * recommend console-tools (closes: #487514)
  * update debian build (closes: #800164)

 -- Radovan Garabík <garabik@kassiopeia.juls.savba.sk>  Tue, 27 Oct 2015 19:13:28 +0100

fonty-rg (0.5) unstable; urgency=high

  * fix typo in README (closes: #163661)
  * capitalize Unicode (closes: #172059)
  * fix typo in manpage (closes: #301890)
  * rename uni to utf8 (closes: #308246). High urgency, since it is a grave bug
  * change my e-mail

 -- Radovan Garabík <garabik@kassiopeia.juls.savba.sk>  Thu, 12 May 2005 10:10:34 +0200

fonty-rg (0.4) unstable; urgency=low

  * install changelog (oops). (closes: #167026)

 -- Radovan Garabík <garabik@melkor.dnp.fmph.uniba.sk>  Tue,  5 Nov 2002 10:49:03 +0100

fonty-rg (0.3) unstable; urgency=low

  * changed Description: (closes: #162085)
  * added ./ in front of build.sh command (closes: #164835)

 -- Radovan Garabik <garabik@melkor.dnp.fmph.uniba.sk>  Tue, 24 Sep 2002 20:04:18 +0200

fonty-rg (0.2) unstable; urgency=low

  * changed Architecture: any into Architecture: all

 -- Radovan Garabik <garabik@melkor.dnp.fmph.uniba.sk>  Thu, 12 Sep 2002 16:10:24 +0200

fonty-rg (0.1) unstable; urgency=low

  * Initial Release (closes: #159010)

 -- Radovan Garabik <garabik@melkor.dnp.fmph.uniba.sk>  Wed, 21 Aug 2002 13:08:38 +0200

Local variables:
mode: debian-changelog
End:


build.sh
#!/bin/sh

export LC_ALL=POSIX

rm -f uni-16.sbf
for i in source/????-????*.sbf
do
  cat $i >>uni-16.sbf
  echo >>uni-16.sbf
  echo >>uni-16.sbf
done

./compact uni-16.sbf > uni-compact-16.sbf
gzip -f -v9 uni-compact-16.sbf


./choose  charsets/LatCyrGr uni-16.sbf   |./compact | ./vga -512 | ./sbf2psf > LatCyrGr-16.psf
gzip -f -v9 LatCyrGr-16.psf

charsets/cz2t.sh charsets/chavo.chars > chavo
cat charsets/graphics >> chavo
echo U+FFFD >> chavo
./choose  chavo uni-16.sbf   | ./compact | ./vga | ./sbf2psf > chavo.psf
gzip -f -v9 chavo.psf
#rm -f chavo

for i in charsets/*.txt;
do
  bn=`basename $i .txt`
  echo Creating font $bn
  charsets/cz2t.sh $i > $bn
  cat charsets/graphics >> $bn
  echo U+FFFD >> $bn
  ./choose $bn uni-16.sbf | ./vga | ./sbf2psf > $bn.psf
  gzip -f -v9 $bn.psf
  rm -f $bn
done

./charset2acm.py charsets/iso8859-16.txt > iso16.acm
gzip -f -v -9 iso16.acm



hex2sbf.py
#!/usr/local/ActivePython-2.1/bin/python

import sys, string

def hex2bin(s):
    return {
    '0': '0000',
    '1': '0001',
    '2': '0010',
    '3': '0011',
    '4': '0100',
    '5': '0101',
    '6': '0110',
    '7': '0111',
    '8': '1000',
    '9': '1001',
    'a': '1010',
    'b': '1011',
    'c': '1100',
    'd': '1101',
    'e': '1110',
    'f': '1111',
    }[string.lower(s)]

import sys
for ll in sys.stdin.xreadlines():
    l = string.strip(ll)
    cp = l[:4]
    glyph = l[5:]
    if len(glyph) == 64: # double width
        continue
        print
        print '[U+%s]' % cp
        for i in range(0, len(glyph), 4):
            cl = hex2bin(glyph[i])+hex2bin(glyph[i+1])+hex2bin(glyph[i+2])+hex2bin(glyph[i+3])
            ccl = []
            for j in range(0, len(cl), 2):
                if '1' in cl[j:j+2]:
                    ccl.append('1')
                else:
                    ccl.append('0')
            cl = string.join(ccl, '')
            cl = string.replace(cl, '0', '.')
            cl = string.replace(cl, '1', '0')
            print cl+'.'
    else:
        print
        print '[U+%s]' % cp
        for i in range(0, len(glyph), 2):
            cl = hex2bin(glyph[i])+hex2bin(glyph[i+1])
            cl = string.replace(cl, '0', '.')
            cl = string.replace(cl, '1', '0')
            print cl+'.'
            
            
            
        

hex2bdf.pl
#!/usr/bin/perl


# There are now four ways to call hex2bdf:
# 1. hex2bdf <unifont.hex >unifont.bdf
# 2. hex2bdf yes <unifont.hex >unifont-double.bdf
# 3. hex2bdf both <unifont.hex >unifont-both.bdf
# 4. hex2bdf shrink <unifont.hex >unifont-shrink.bdf

# 1. & 2. are as in the original hex2bdf.

# 3. includes both single and double width characters in a single font
# (since some applications prefer this, e.g. GNU emacs 21 work fine with
# this). It was done like that in the older GNU unifont versions.

# And 4. generates a font, which has all 16-wide characters shrunken to
# 8-wide ones (for cell-spaced font application like konsole).



$double = shift(ARGV);
if ($double =~ /^(y(es)?|t(rue)|1)/i) {
    $double = 'yes';
} elsif ($double =~ /^s(hrink)?/i) {
    $double = 'shrink';
} elsif ($double =~ /^b(oth)?/i) {
    $double = 'both';
} elsif (($double =~ /^no?/i) || ($double eq '')) {
    $double = 'no';
} else {
    warn "unknown double specifier '$double'";
    $double = 'no';
}

if ($double eq 'yes') {
  while (<>) { $glyph{$1} = $2 if /(....):([0-9A-Fa-f]{64})\n/; }
  $boxwidth = 16;
  $halffull = 'f';
        $spacing = 'c';
  $defaultchar = 0xFF0A; # FULLWIDTH ASTERISK
}
elsif ($double eq 'no') {
  while (<>) { $glyph{$1} = $2 if /(....):([0-9A-Fa-f]{32})\n/; }
  $boxwidth = 8;
  $halffull = 'h';
        $spacing = 'c';
  $defaultchar = 0xFFFD;
} elsif ($double eq 'shrink') {
  for (my $i=0; $i<256; $i++) {
                my $t = (($i&0x80)?0x8:0) | (($i&0x40)?0x8:0) |
                                    (($i&0x20)?0x4:0) | (($i&0x10)?0x4:0) |
                                                (($i&0x08)?0x2:0) | (($i&0x04)?0x2:0) |
                                                (($i&0x02)?0x1:0) | (($i&0x01)?0x1:0);
                $narrow{sprintf "%02X", $i} = sprintf "%X", $t;
  };
  while (<>) {
                        next unless /(....):([0-9A-Fa-f]{32,64})\n/;
                        my $i = $1;
      my $g = $2;
                        $g =~ s/(..)/$narrow{$&}/ge if length($g) eq 64;
                        $glyph{$i} = $g;
  }
  $boxwidth = 8;
  $halffull = '';
        $spacing = 'c';
  $defaultchar = 0xFFFD;
  $namesuffix = '.shrink';
} elsif ($double eq 'both') {
  while (<>) { $glyph{$1} = $2 if /(....):([0-9A-Fa-f]{32,64})\n/; }
  $boxwidth = 16;
  $halffull = '';
        $spacing = 'p';
        $namesuffix = '.twowidth';
  $defaultchar = 0xFF0A; # FULLWIDTH ASTERISK
};

@chars = sort keys %glyph; $[ = 1;
# dbmopen (%charname, "/usr/share/unicode/unidata/charname.db", 0);

print <<EOF;
STARTFONT 2.1
FONT
-gnu-fixed$namesuffix-medium-r-normal-$halffull-$boxwidth-160-75-75-$spacing-80-\
iso10646-1
SIZE 16 75 75
FONTBOUNDINGBOX $boxwidth 16 0 -2
STARTPROPERTIES 18
FONTNAME_REGISTRY ""
FOUNDRY "gnu"
FAMILY_NAME "fixed"
WEIGHT_NAME "medium"
SLANT "r"
SETWIDTH_NAME "normal"
ADD_STYLE_NAME "$halffull"
PIXEL_SIZE 16
POINT_SIZE 160
RESOLUTION_X 75
RESOLUTION_Y 75
SPACING "c"
AVERAGE_WIDTH 80
CHARSET_REGISTRY "iso10646"
CHARSET_ENCODING "1"
DEFAULT_CHAR $defaultchar
FONT_ASCENT 14
FONT_DESCENT 2
ENDPROPERTIES
CHARS $#chars
EOF

foreach $character (@chars)
{
        $encoding = hex($character); $glyph = $glyph{$character};
        $width = length ($glyph) > 32 ? 2 : 1;
        $dwidth = $width * 8; $swidth= $width * 500;
        $glyph =~ s/((..){$width})/\n$1/g;
        $character = "$character $charname"
            if $charname = $charname{pack("n",hex($character))};

        print "STARTCHAR U+$character
ENCODING $encoding
SWIDTH $swidth 0
DWIDTH $dwidth 0
BBX $dwidth 16 0 -2
BITMAP $glyph
ENDCHAR\n";
}

print "ENDFONT\n";

